Remove platform-specific translations support

vadz opened the pull request:
This seems like a better way to do it and it allows to get rid of unnecessary complexity associated with having these platform-specific message catalogs.

@lanurmi @vslavik please let me know if you have any comments about this, TIA!

vadz committed:
Check stream error in LastRead() unit test too

It doesn't make sense to call LastRead() if an error happened while
reading, so check if it did.

vadz committed:
Factor out platform detection from XRC code into wxPlatformId

This will allow reusing the same strings in other places.

Also use this as opportunity to add "msw" as the (preferred) synonym for
"win" for general consistency.

vadz committed:
Remove platform-specific translations support

As shown by the fact that no other platform-specific translations were
ever added since the initial version of locale/msw/it.po added way back
in 4d931bcca0 (Translate '&Help' to '&' for italian Windows only,
2005-08-12), this is not especially useful, so just handle this
particular case specially in wxGetStockLabel() and remove support for
platform-specific translations.

This allows to simplify the makefiles and the catalog loading code as
there is no need to deal with the files in subdirectories any more.

vadz committed:
Use wxGetStockLabel() in internat sample

This allows to reuse "&File" translation already available in wxstd
catalog rather than translating it again in the sample catalogs for all
languages, so remove the now unnecessary translations from them.

vadz committed:
Add an almost empty message catalog with Georgian translations

Surprisingly, we had Georgian translations for the internat sample, but
not for wx itself.

Extract a couple of strings that were present in the sample catalog for
the string defined in wx into the main catalog for tidiness.

lanurmi commented on interface/wx/translation.h:
```diff
@@ -476,6 +476,43 @@ class wxMsgCatalog
+
+    @since 3.1.5
+ */
+#define wxGETTEXT_IN_PLATFORM_CONTEXT(context, string)
```
There is a big problem with this approach: `xgettext` will only see the literal value of the `context` argument. If `xgettext` was run on the code sample above, it would produce an entry with `msgctxt "menu"`, yet at runtime an entry with context `"msw.menu"` would be searched for, and that wouldn't be a match.

vadz replied:
I don't see how could this be avoided. The (very) few cases when platform-specific translations are needed would have to be handled manually, i.e. by editing the context in the `.po` file.

What else could we possibly do?

lanurmi replied:
Any manual modifications to `.po` files will be commented out by the next run of `msgmerge`. So the modifications would need to be done to the `wxstd.pot` file really. This is doable within wxWidgets's own translations, as part of `wxstd.pot` generation -- but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.

From the perspective of library user this manual editing of `.po` or `.pot` is too complicated, IMHO. It would require an unusual additional step to the `xgettext` extraction process, and if omitted, there would be no error at any point, but messages with platform-specific context would remain untranslated for no obvious reason. Again, explicitly listing the three possible platforms could be a solution more easy to grasp.

vadz replied:
> Any manual modifications to `.po` files will be commented out by the next run of `msgmerge`.

Oops, yes, you're right, this is indeed a problem, thanks for pointing it out.

> So the modifications would need to be done to the `wxstd.pot` file really. This is doable within wxWidgets's own translations, as part of `wxstd.pot` generation -- but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.

Sorry, I'm not sure about what exactly do you propose. Could you please show how would this work?

I.e. we could add `wxTRANSLATE_IN_CONTEXT()`, but how would it be used? We can't write
```cpp
wxTRANSLATE_IN_CONTEXT("mac", "foo")
wxTRANSLATE_IN_CONTEXT("msw", "foo")
wxTRANSLATE_IN_CONTEXT("unix", "foo")
```
for each string...

> From the perspective of library user this manual editing of `.po` or `.pot` is too complicated, IMHO.

Yes, I have to agree with this.

Finally, it seems using contexts for this might not have been a good idea. At least it definitely would be much simpler to just make a special exception for `wxID_HELP` in `wxGetStockLabel()` rather than trying to handle this at translations level.

Or do you (or anybody else) still see some way to make the idea of using contexts work?

lanurmi replied:
> > but on the other hand, if indeed there are very few cases within wxWidgets itself, wouldn't it be simpler to just explicitly call the `wxGETTEXT_IN_CONTEXT` macro once for each three platforms? Or call a context-aware no-op macro similar to `wxTRANSLATE`.
> 
> Sorry, I'm not sure about what exactly do you propose. Could you please show how would this work?
> 
> I.e. we could add `wxTRANSLATE_IN_CONTEXT()`, but how would it be used? We can't write
> 
> ```c++
> wxTRANSLATE_IN_CONTEXT("mac", "foo")
> wxTRANSLATE_IN_CONTEXT("msw", "foo")
> wxTRANSLATE_IN_CONTEXT("unix", "foo")
> ```
> 
> for each string...

That's pretty much what I was proposing, and whether such can be written for each string, depends on how many platform-specific strings there are in the first place. So far there has been exactly one, the `"&Help"` one. It wouldn't be unrealistic to write those three lines for `"&Help"` _once_ in the wx source code. Keeping in mind that those do not need to be repeated with every call to `wxGETTEXT_IN_PLATFORM_CONTEXT`, just have them once, somewhere.

For a handful strings maintaining such lists manually would still be feasible...

> Finally, it seems using contexts for this might not have been a good idea. At least it definitely would be much simpler to just make a special exception for `wxID_HELP` in `wxGetStockLabel()` rather than trying to handle this at translations level.

This might be true with strings in the generic or common parts of the source. But in e.g. `src/osx/cocoa/menu.mm` and `src/osx/menu_osx.cpp` there are many platform-specific strings that could be contextized using just the existing macros, as their platform is fixed.

> Or do you (or anybody else) still see some way to make the idea of using contexts work?

I'll keep thinking about options...

vadz replied:
My idea was to automatically pick up the translations in `$platform.menu` context for the stock labels if they are present. Which already doesn't seem really practical, i.e. we're not going to have `2*3*72` string (see `wxGetStockLabel()`).

It's true that using contexts could still be useful for the strings in platform-specific code, but then we don't need to use `wxPlatformId` and can just hardcode the context, so I still don't see any real use for `wxGETTEXT_IN_PLATFORM_CONTEXT()`...

vslavik replied:
Thanks @lanurmi for saying what I would say, I agree 100% with everything.

`#ifdef` with an explanatory `// TRANSLATORS:` comment seems appropriate for handling of the *Help* string — it's unlikely anything else of this kind will be required.

vadz replied:
@vslavik Do you also think that `wxGETTEXT_IN_PLATFORM_CONTEXT` can still be useful? AFAICS if there is no way to use it with the standard gettext workflow without hacks, it's just not useful at all, so I'll just drop it, unless I'm missing something.

vslavik replied:
> @vslavik Do you also think that wxGETTEXT_IN_PLATFORM_CONTEXT can still be useful?

I got a race condition with that one. For just one string, used from one place, it's not useful.

It would be potentially useful if we had a few strings like "Help" and/or present in several places in the code. Then you could use `wxGETTEXT_IN_PLATFORM_CONTEXT` at the place of use, avoiding repetitive `#ifdef` logic, and have just one "dummy" place in code where all strings would be manually listed for `xgettext` extraction. It's a similar situation to `wxGetTranslation` (runtime invocation without a string literal) and `wxTRANSLATE` (some placement in code to support extraction).

IOW, I'd say it's something that might be useful in the future (but probably not), but not worth keeping now (not least because it would confuse readers who'd assume it's *needed* for something).

discnl reviewed (commented):
Looks good, but not familiar enough with this so I won't give it a then phony seal of approval. Besides minor review I at least did a simple check on macOS to see the internat sample still translating the menus (particularly `File`).

vadz commented on include/wx/translation.h:
```diff
@@ -57,6 +59,12 @@
     wxGetTranslation((sing), (plur), n, wxString(), c)
 
+#define wxGETTEXT_PLATFORM_CONTEXT(c) \
+    (wxPlatformId::GetCurrent() + '.' + c)
```
`wxS()` is not needed (and, in fact, wouldn't work) around _character_ (as opposed to string) constants.

discnl replied:
Perhaps not here then (a line which may get deleted anyway...) with some string config I haven't tried. For me with `wxS()` it did at least compile before. I checked with MSW now: `wxString`'s `wchar_t, size_t nRepeat` ctor is used, otherwise `char, size_t`. Also there's  quite some `wxS('` hits, all in library code, and I've been using it myself for some time.

vadz commented on samples/internat/internat.cpp:
```diff
@@ -327,10 +326,17 @@ MyFrame::MyFrame(wxLocale& locale)
     macro_menu->Append(INTERNAT_MACRO_8, wxGETTEXT_IN_CONTEXT_PLURAL("context_2", "sing", "plur", 1));
     macro_menu->Append(INTERNAT_MACRO_9, wxGETTEXT_IN_CONTEXT_PLURAL("context_2", "sing", "plur", 2));
 
+    wxMenu *help_menu = new wxMenu;
```
I typically try to keep the style locally consistent first, before making it globally consistent, so I'd rather keep it like this.

I'm still waiting for clang-format to become good enough to run it on our code base to harmonize the style of it once and for all...

discnl replied:
Thanks, I'll stick to local consistency then. I do like the incremental changes: as a reader they can give me small hints which code is newer. The inconsistency bothers me less in such cases. Also it means a few drops of blame noise less with a global format replacement ;).

> I'm still waiting for clang-format to become good enough to run it on our code base to harmonize the style of it once and for all...

I haven't used it, but I thought it was 'just' a matter of writing a good enough formatting file. Or it's already known to not be adequate for wx?

vadz replied:
> > I'm still waiting for clang-format to become good enough to run it on our code base to harmonize the style of it once and for all...
> 
> I haven't used it, but I thought it was 'just' a matter of writing a good enough formatting file. Or it's already known to not be adequate for wx?

Last time I checked, it didn't support quite the style we use, so running it would result in an even greater amount of changes that it normally does. Maybe we should still do it though, it's true that fixing style all the time is tiresome.

vadz commented on src/xrc/xmlres.cpp:
```diff
@@ -42,6 +42,7 @@
 #include "wx/hashset.h"
 #include "wx/scopedptr.h"
 #include "wx/config.h"
+#include "wx/platinfo.h"
```
This is another "local style" thing: I definitely prefer to sort the headers alphabetically, but if they're not already sorted (and here they aren't), I just add the new ones at the end.

Ideal would be to sort all of them once and then add a style check that they remain sorted.

vadz committed:
Recognize wxGETTEXT_IN_CONTEXT macros in wx sources

Tell xgettext to extract the strings appearing inside
wxGETTEXT_IN_CONTEXT and wxGETTEXT_IN_CONTEXT_PLURAL too.

vadz commented:
So here is the much simpler solution using just a special hack for this single label.

vslavik commented on docs/doxygen/overviews/xrc_format.h:
```diff
@@ -2674,14 +2674,15 @@ should be processed on. It is filtered out and ignored on any other platforms.
 @beginDefList
-@itemdef{ @c win, Windows }
-@itemdef{ @c mac, macOS (or Mac Classic in wxWidgets version supporting it) }
+@itemdef{ @c msw, Windows, preferred platform name }
+@itemdef{ @c win, Windows, alternative synonym }
```
FWIW, I think "msw" is strictly worse than "win" as the platform identifier. Unlike "win", it is not universally recognized as an abbrev. for Windows (see http://google.com/search?q=msw; I don't think I ever saw it in non-wx context…). This value is the *platform* ("Windows", "Mac"), not wx *port* ("wxMSW").

vadz replied:
But we do use MSW a lot and quite consistently. I'm also not completely unsympathetic to people who follow FSF advice not to use "win" as abbreviation of "Windows", even if I don't see any real problem with this myself, and using "msw" allows to completely side-step the issue.

vslavik reviewed (approved):
Thanks, I think this is a better approach. I have some minor reservations to the details, but that's nitpicking.

vadz commented on src/common/stockitem.cpp:
```diff
@@ -120,6 +120,15 @@ wxString wxGetStockLabel(wxWindowID id, long flags)
+    // translation for some languages (Italian), so try this first.
+    if ( id == wxID_HELP && (flags & wxSTOCK_WITH_MNEMONIC) )
+    {
+        stockLabel = wxGETTEXT_IN_CONTEXT("msw.menu", "&Help");
```
Good point, will change to "standard Windows menu", thanks.
